File: app/src/main/java/com/hiddenramblings/tagmo/amiibo/AmiiboFile.java
package com.hiddenramblings.tagmo.amiibo;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.documentfile.provider.DocumentFile;

import java.io.File;

public class AmiiboFile implements Parcelable {

    protected File filePath;
    protected DocumentFile docPath;
    protected long id;
    protected byte[] data;

    public AmiiboFile(File filePath, long id, byte[] data) {
        this.filePath = filePath;
        this.id = id;
        this.data = data;
    }

    public AmiiboFile(File filePath, long id) {
        this(filePath, id, null);
    }

    public AmiiboFile(DocumentFile docPath, long id, byte[] data) {
        this.docPath = docPath;
        this.id = id;
        this.data = data;
    }

    public byte[] getData() {
        return data;
    }

    public void setData(byte[] data) {
        this.data = data;
    }

    public File getFilePath() {
        return filePath;
    }

    public DocumentFile getDocUri() {
        return docPath;
    }

    @SuppressWarnings("unused")
    public void setFilePath(File filePath) {
        this.filePath = filePath;
    }

    @SuppressWarnings("unused")
    public void setDocUri(DocumentFile docPath) {
        this.docPath = docPath;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeSerializable(this.filePath);
        dest.writeLong(this.id);
        dest.writeInt(this.data.length);
        dest.writeByteArray(this.data);
    }

    protected AmiiboFile(Parcel in) {
        this.filePath = (File) in.readSerializable();
        this.id = in.readLong();
        this.data = new byte[in.readInt()];
        in.readByteArray(this.data);
    }

    public static final Parcelable.Creator<AmiiboFile> CREATOR = new Parcelable.Creator<>() {
        @Override
        public AmiiboFile createFromParcel(Parcel source) {
            return new AmiiboFile(source);
        }

        @Override
        public AmiiboFile[] newArray(int size) {
            return new AmiiboFile[size];
        }
    };
}

File: app/src/main/java/com/hiddenramblings/tagmo/amiibo/AmiiboFile.kt
package com.hiddenramblings.tagmo.amiibo

import android.os.Parcel
import android.os.Parcelable
import androidx.documentfile.provider.DocumentFile
import com.hiddenramblings.tagmo.amiibo.tagdata.AmiiboData
import com.hiddenramblings.tagmo.eightbit.io.Debug
import com.hiddenramblings.tagmo.eightbit.os.Version
import com.hiddenramblings.tagmo.nfctech.Foomiibo
import com.hiddenramblings.tagmo.nfctech.TagArray
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import java.io.File

open class AmiiboFile : Parcelable {
    var filePath: File? = null
    var docUri: DocumentFile? = null
    var id: Long
    var data: ByteArray?

    @JvmOverloads
    constructor(filePath: File?, id: Long, data: ByteArray? = null) {
        this.filePath = filePath
        this.id = id
        this.data = data
    }

    constructor(docPath: DocumentFile?, id: Long, data: ByteArray?) {
        docUri = docPath
        this.id = id
        this.data = data
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeSerializable(filePath)
        dest.writeLong(id)
        data?.let {
            dest.writeInt(it.size)
            dest.writeByteArray(it)
        }
    }

    protected constructor(parcel: Parcel) {
        filePath = if (Version.isTiramisu)
            parcel.readSerializable(null, File::class.java)
        else
            @Suppress("deprecation") parcel.readSerializable() as File?
        id = parcel.readLong()
        data = ByteArray(parcel.readInt()).also { parcel.readByteArray(it) }
    }

    suspend fun withRandomSerials(keyManager: KeyManager, count: Int) : ArrayList<AmiiboData> {
        val dataList: ArrayList<AmiiboData> = arrayListOf()
        coroutineScope {
            val tagData = TagArray.getValidatedAmiibo(keyManager, this@AmiiboFile)
            (0 until count).map {
                async(Dispatchers.IO) {
                    try {
                        AmiiboData(keyManager.decrypt(tagData)).apply {
                            uID = Foomiibo.generateRandomUID()
                        }.also {
                            dataList.add(it)
                        }
                    } catch (e: Exception) {
                        Debug.warn(e)
                    }
                }
            }.awaitAll()
        }
        return dataList
    }

    suspend fun withRandomSerials(count: Int) : ArrayList<AmiiboData> {
        val dataList: ArrayList<AmiiboData> = arrayListOf()
        this@AmiiboFile.data?.let { tagData ->
            coroutineScope {
                (0 until count).map {
                    async(Dispatchers.IO) {
                        try {
                            AmiiboData(tagData).apply {
                                uID = Foomiibo.generateRandomUID()
                            }.also {
                                dataList.add(it)
                            }
                        } catch (e: Exception) {
                            Debug.warn(e)
                        }
                    }
                }.awaitAll()
            }
        }
        return dataList
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<AmiiboFile?> = object : Parcelable.Creator<AmiiboFile?> {
            override fun createFromParcel(source: Parcel): AmiiboFile {
                return AmiiboFile(source)
            }

            override fun newArray(size: Int): Array<AmiiboFile?> {
                return arrayOfNulls(size)
            }
        }
    }
}